R/eval-walk.R

```diff
@@ -172,18 +175,26 @@ walk_data_tree <- function(expr, data_mask, context_mask, colon = FALSE) {
   vars <- data_mask$.__tidyselect__.$internal$vars
   strict <- data_mask$.__tidyselect__.$internal$strict
   data <- data_mask$.__tidyselect__.$internal$data
+  allow_predicates <- data_mask$.__tidyselect__.$internal$allow_predicates
 
   as_indices_sel_impl(
     out,
     vars = vars,
     strict = strict,
     data = data,
+    allow_predicates = allow_predicates,
     call = error_call
   )
 }
 
-as_indices_sel_impl <- function(x, vars, strict, data = NULL, call) {
+as_indices_sel_impl <- function(x, vars, strict, data = NULL, allow_predicates = TRUE, call) {
   if (is.function(x)) {
+    if (!allow_predicates) {
+      cli::cli_abort(
```

Comment:
The class idea is good, but feeding through a type summary of the original object is going to be hard because we'd need to capture it before proxying it and pass it down through many levels.